process/interceptors/factory/eligibleNodesCache.go

```diff
@@ -0,0 +1,243 @@
+package factory
+
+import (
+	"sync"
+
+	"github.com/multiversx/mx-chain-core-go/core"
+	"github.com/multiversx/mx-chain-core-go/core/check"
+	"github.com/multiversx/mx-chain-go/process"
+	logger "github.com/multiversx/mx-chain-logger-go"
+)
+
+var log = logger.GetOrCreate("eligibleNodesCache")
+
+// max number of epochs to keep in the local map
+const epochsDelta = uint32(3)
+
+type epochLimits struct {
+	min uint32
+	max uint32
+}
+
+func (limits *epochLimits) isHigherInRange(newEpoch uint32) bool {
+	return newEpoch > limits.min && newEpoch < limits.min+epochsDelta
+}
+
+func (limits *epochLimits) isHigherOutOfRange(newEpoch uint32) bool {
+	return newEpoch >= limits.min+epochsDelta
+}
+
+func (limits *epochLimits) isLowerInRange(newEpoch uint32) bool {
+	return newEpoch < limits.max && newEpoch > limits.max-epochsDelta && limits.max >= epochsDelta
+}
+
+func (limits *epochLimits) isLowerOutOfRange(newEpoch uint32) bool {
+	return newEpoch <= limits.max-epochsDelta && limits.max >= epochsDelta
+}
+
+type eligibleNodesCache struct {
+	// eligibleNodesMap holds a map with key shard id and value
+	//		a map with key epoch and value
+	//		a map with key eligible node pk and value struct{}
+	// this is needed because:
+	// 	- a shard node will intercept proof for self shard and meta
+	// 	- a meta node will intercept proofs from all shards
+	eligibleNodesMap map[uint32]map[uint32]map[string]struct{}
+	// epochsLimitsForShardMap holds map with key shard id and value the limits for that shard
+	epochsLimitsForShardMap map[uint32]*epochLimits
```

Comment:
we can have a separate component for `eligibleNodesMap` per shard, coupled with `epochsLimitsForShardMap`, because we already do most of the operations per shard

Reply:
but it can be also considered later on with a TODO

Reply:
added todo